src/services/providers/provider-sources/load-balance-provider.ts

```diff
@@ -0,0 +1,164 @@
+          try {
+            const response = await transport.request({ method, params });
+            rpcMetrics[index][method] = updateMetrics(start, metrics, true);
+            return response;
+          } catch (error) {
+            // If the error is a rejection, we should throw it
+            if (shouldThrow(error as Error)) throw error;
+
+            rpcMetrics[index][method] = updateMetrics(start, metrics, false);
+            availableTransports.shift(); // Remove this transport. The best is the first, so we can shift
+            return processRequest(); // Try the next best transport
+          }
+        };
+
+        return processRequest();
+      },
+      type: 'load-balance',
+    });
+  };
+}
+
+function initializeMetrics(): RPCMetrics {
+  return {
+    avgProcessingTime: 0,
+    successRate: 1,
+    pending: 0,
+    samples: [],
+    successCount: 0, // We need this to calculate the success rate when removing samples
+  };
+}
+
+function removeInvalidSamples(metrics: RPCMetrics, currentTime: number, ttl: TimeString): RPCMetrics {
+  const cutoffTime = currentTime - ms(ttl);
+  const validSampleIndex = metrics.samples.findIndex((time) => time >= cutoffTime);
+
+  if (validSampleIndex === -1) {
+    return initializeMetrics();
+  } else if (validSampleIndex > 0) {
+    const removedSamples = validSampleIndex;
+    metrics.successCount = Math.round(metrics.successRate * (metrics.samples.length - removedSamples));
+    metrics.samples = metrics.samples.slice(validSampleIndex);
+    metrics.successRate = metrics.successCount / metrics.samples.length;
+  }
+
+  return metrics;
+}
+
+function updateMetrics(start: number, metrics: RPCMetrics, isSuccess: boolean): RPCMetrics {
+  const processingTime = Date.now() - start;
+  const newSampleCount = metrics.samples.length + 1;
+  metrics.avgProcessingTime = (metrics.avgProcessingTime * metrics.samples.length + processingTime) / newSampleCount;
+  metrics.successCount += isSuccess ? 1 : 0;
+  metrics.successRate = metrics.successCount / newSampleCount;
+  metrics.pending--;
+  metrics.samples.push(Date.now());
+  return metrics;
+}
+
+function calculateScore(metrics: RPCMetrics) {
+  return metrics.successRate * 0.7 + (1 - metrics.avgProcessingTime / 1000) * 0.3;
+}
+
+function shouldThrow(error: Error) {
+  if ('code' in error && typeof error.code === 'number') {
+    if (
+      error.code === TransactionRejectedRpcError.code ||
+      error.code === UserRejectedRequestError.code ||
+      error.code === 5000 // CAIP UserRejectedRequestError
+    )
+      return true;
+  }
+  return false;
+}
+
+export type LoadBalanceConfig = {
+  key?: TransportConfig['key'];
+  name?: TransportConfig['name'];
+  minSuccessRate?: number;
+  minSamples?: number;
+  maxAttempts?: number;
+  samplesTtl?: TimeString;
+};
+
+type RPCMetrics = {
+  avgProcessingTime: number;
+  successRate: number;
+  pending: number;
+  samples: number[];
+  successCount: number;
+};
```

Comment:
I think everything would be simpler if you just do something like this:
```suggestion
  pending: number;
  samples: { timestamp: Timestamp, success: boolean, processingTime: number} [];
};
```
Then you can filter out old samples and calculate averages correctly